networking/nat/src/main/java/tech/pegasys/teku/networking/nat/NatManager.java

```diff
@@ -0,0 +1,418 @@
+import org.jupnp.support.igd.callback.GetExternalIP;
+import org.jupnp.support.igd.callback.PortMappingAdd;
+import org.jupnp.support.igd.callback.PortMappingDelete;
+import org.jupnp.support.model.PortMapping;
+import tech.pegasys.teku.infrastructure.async.FutureUtil;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.service.serviceutils.Service;
+
+@SuppressWarnings("unchecked")
+public class NatManager extends Service {
+  protected static final Logger LOG = LogManager.getLogger();
+
+  static final String SERVICE_TYPE_WAN_IP_CONNECTION = "WANIPConnection";
+
+  private final UpnpService upnpService;
+  private final RegistryListener registryListener;
+  private final SafeFuture<String> externalIpQueryFuture = new SafeFuture<>();
+
+  private final Map<String, SafeFuture<RemoteService>> recognizedServices = new HashMap<>();
+  private volatile Queue<NatPortMapping> forwardedPorts = new ConcurrentLinkedQueue<>();
+  private Optional<String> discoveredOnLocalAddress = Optional.empty();
+
+  public NatManager(final NatMethod natMethod) {
+    // this(new UpnpServiceImpl(new DefaultUpnpServiceConfiguration()));
+
+    // Workaround for an issue in the jupnp library: the ExecutorService used misconfigures
+    // its ThreadPoolExecutor, causing it to only launch a single thread. This prevents any work
+    // from getting done (effectively a deadlock). The issue is fixed here:
+    //   https://github.com/jupnp/jupnp/pull/117
+    // However, this fix has not made it into any releases yet.
+    // TODO: once a new release is available, remove this @Override
+    this(new UpnpServiceImpl(new TekuNatServiceConfiguration()), natMethod);
+  }
+
+  /**
+   * Constructor
+   *
+   * @param service is the desired instance of UpnpService.
+   */
+  NatManager(final UpnpService service, final NatMethod natMethod) {
+    upnpService = service;
+
+    // prime our recognizedServices map so we can use its key-set later
+    recognizedServices.put(SERVICE_TYPE_WAN_IP_CONNECTION, new SafeFuture<>());
+
+    // registry listener to observe new devices and look for specific services
+    registryListener =
+        new TekuRegistryListener() {
+          @Override
+          public void remoteDeviceAdded(final Registry registry, final RemoteDevice device) {
+            LOG.debug("UPnP Device discovered: " + device.getDetails().getFriendlyName());
+            inspectDeviceRecursive(device, recognizedServices.keySet());
+          }
+        };
+  }
+
+  private void inspectDeviceRecursive(final RemoteDevice device, final Set<String> serviceTypes) {
+    for (RemoteService service : device.getServices()) {
+      String serviceType = service.getServiceType().getType();
+      if (serviceTypes.contains(serviceType)) {
+        synchronized (this) {
+          // log a warning if we detect a second WANIPConnection service
+          SafeFuture<RemoteService> existingFuture = recognizedServices.get(serviceType);
```

Comment:
We aren't being consistent in the use of `synchronised` when accessing `recognizedServices`.  The `serviceTypes` set is actually the keyset of `recognizedServices` but is being accessed outside the `synchronized` block which isn't thread safe.  It's also accessed in the constructor without being synchronized.

I'd suggest removing the `synchronized` and just making `recognizedServices` a `ConcurrentHashMap`.  This little section then becomes:
```suggestion
      SafeFuture<RemoteService> existingFuture = recognizedServices.get(serviceType);
      if (existingFuture != null) {
          // log a warning if we detect a second WANIPConnection service
```

That said, the warning logged here assumes we only care about one service, in which case we could just remove `recognizedServices` and have a single field that holds the relevant `SafeFuture`.